src/accumulation.jl

```diff
@@ -3,13 +3,80 @@
 
 Returns `x+y`, potentially mutating `x` in-place to hold this value.
 This avoids allocations when `x` can be mutated in this way.
-
-See also: [`InplaceableThunk`](@ref).
 """
 add!!(x, y) = x + y
 
-add!!(x, t::InplaceableThunk) = t.add!(x)
+"""
+    add!!(x, t::ImplacableThunk)
+
+The specialization of `add!!` for [`InplaceableThunk`](@ref) promises to only call
+`t.add!` on `x` if `x` is suitably mutable; otherwise it will be out of place.
+"""
+function add!!(x, t::InplaceableThunk)
+    return if is_inplaceable_destination(x)
+        if !debug_mode()
+            t.add!(x)
+        else
+            return debug_add!(x, t)
+        end
+    else
+        x + t
+    end
+end
+
+function add!!(x::AbstractArray{<:Any, N}, y::AbstractArray{<:Any, N}) where N
+    return if is_inplaceable_destination(x)
+        x .+= y
+    else
+        x + y
+    end
+end
+
+
+"""
+    is_inplaceable_destination(x)
+
+Returns true if `x` is suitable for for storing inplace accumulation of gradients.
+For arrays this boils down `x .= y` if will work to mutate `x`, if `y` is an appropriate
+differential.
+Wrapper array types do not need to overload this if they overload `Base.parent`, and are
+`is_inplaceable_destination` if and only if their parent array is.
+Other types should overload this, as it defaults to `false`.
+"""
+is_inplaceable_destination(::Any) = false
+is_inplaceable_destination(::Array) = true
+is_inplaceable_destination(::SparseVector) = true
+is_inplaceable_destination(::SparseMatrixCSC) = true
+is_inplaceable_destination(::BitArray) = true
+function is_inplaceable_destination(x::AbstractArray)
+    p = parent(x)
+    p === x && return false  # no parent
+    # basically all wrapper types delegate `setindex!` to their `parent` after some
+    # processing and so are mutable if their `parent` is.
+    return is_inplaceable_destination(p)
+end
+
+function debug_add!(accumuland, t::InplaceableThunk)
+    returned_value = t.add!(accumuland)
+    if returned_value !== accumuland
```

Comment:
Perhaps this is cheap enough to do it outside of `debug_mode`?
Maybe if we find outselves having to tell people to turn on `debug_mode` to find errors we should look into that?
Or maybe we should look into it now.
What do you think?